### PR TITLE
Crypto

### PR DESCRIPTION
**Pull request by MykolaSolyanko**

### CODE REVIEW

File: include/aos/crypto/mbedtls/drivers/aos/driver_wrap.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
+/*
```
Comment (al1img) [inline include/aos/crypto/mbedtls/drivers/aos/driver_wrap.h:1]: Should be `hpp` file I guess. Also, rename to `driverwrap.hpp` as we do not use `_` in file names.

File: include/aos/crypto/mbedtls/drivers/aos/driver_wrap.h [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+#ifndef DRIVER_WRAP_H_
+#define DRIVER_WRAP_H_
+
+#if defined(PSA_CRYPTO_DRIVER_AOS)
+
+#include <aos/common/crypto.hpp>
+
+#include "driver.h"
+
+struct KeyDescription {
```
Comment (al1img) [inline include/aos/crypto/mbedtls/drivers/aos/driver_wrap.h:17]: All public definitions should have doxygen comments.

File: include/aos/crypto/mbedtls/drivers/aos/driver_wrap.h [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ * Copyright (C) 2024 Renesas Electronics Corporation.
+ * Copyright (C) 2024 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef DRIVER_WRAP_H_
+#define DRIVER_WRAP_H_
+
+#if defined(PSA_CRYPTO_DRIVER_AOS)
```
Comment (al1img) [inline include/aos/crypto/mbedtls/drivers/aos/driver_wrap.h:11]: I don't think we need this define.

File: src/crypto/mbedtls/drivers/aos/driver_wrap.cpp [hunk 4]
```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright (C) 2024 Renesas Electronics Corporation.
+ * Copyright (C) 2024 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#if defined(PSA_CRYPTO_DRIVER_AOS)
+#include "mbedtls/pk.h"
```
Comment (al1img) [inline src/crypto/mbedtls/drivers/aos/driver_wrap.cpp:9]: <>

File: src/crypto/mbedtls/drivers/aos/driver_wrap.cpp [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+#include "mbedtls/rsa.h"
```
Comment (al1img) [inline src/crypto/mbedtls/drivers/aos/driver_wrap.cpp:10]: ditto

File: src/crypto/mbedtls/drivers/aos/driver_wrap.cpp [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+#include <mbedtls/asn1write.h>
+#include <mbedtls/base64.h>
+#include <mbedtls/pem.h>
+
+#include "aos/common/tools/array.hpp"
+
```
Comment (al1img) [inline src/crypto/mbedtls/drivers/aos/driver_wrap.cpp:16]: remove empty line

File: src/crypto/mbedtls/drivers/aos/driver_wrap.cpp [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+#include "aos/crypto/mbedtls/drivers/aos/driver_wrap.h"
+
```
Comment (al1img) [inline src/crypto/mbedtls/drivers/aos/driver_wrap.cpp:18]: Add static, public headers:
```c++
/***************
 * Static
 ***************/
```

File: src/crypto/mbedtls/drivers/aos/driver_wrap.cpp [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+static aos::StaticArray<KeyDescription, MBEDTLS_PSA_KEY_SLOT_COUNT> builtinKeys;
```
Comment (al1img) [inline src/crypto/mbedtls/drivers/aos/driver_wrap.cpp:19]: use `s` prefix for static var: `sbuiltinKeys`

File: src/crypto/mbedtls/drivers/aos/driver_wrap.cpp [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+            builtinKeys.Remove(&key);
+
+            psa_destroy_key(MBEDTLS_SVC_KEY_ID_GET_KEY_ID(keyId));
+
+            break;
+        }
+    }
+}
+
+#endif /* PSA_CRYPTO_DRIVER_AOS */
```
Comment (al1img) [inline src/crypto/mbedtls/drivers/aos/driver_wrap.cpp:229]: New line at the end.

File: include/aos/crypto/crypto.hpp [hunk 10]
```diff
@@ -0,0 +1,92 @@
+/*
```
Comment (al1img) [inline include/aos/crypto/crypto.hpp:1]: This implementation should be in `aos/common/crypto/mbedtls/crypto.hpp`

File: src/crypto/crypto.cpp [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+#include <psa/crypto.h>
+
+#include "aos/crypto/crypto.hpp"
+#include "aos/crypto/mbedtls/drivers/aos/driver_wrap.h"
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+aos::Error Crypto::PEMToX509Certs(
```
Comment (al1img) [inline src/crypto/crypto.cpp:118]: Location 'src/common/crypto/mbedtls/crypto.cpp`

File: include/aos/crypto/mbedtls/drivers/aos/driver_wrap.h [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+#include "driver.h"
+
+struct KeyDescription {
+    psa_key_id_t                      mKeyId;
+    psa_key_lifetime_t                mLifetime;
+    psa_drv_slot_number_t             mSlotNumber;
+    const aos::crypto::PrivateKeyItf* mPrivKey;
+};
+
+aos::Error AddKey(KeyDescription& key);
```
Comment (al1img) [inline include/aos/crypto/mbedtls/drivers/aos/driver_wrap.h:24]: Rename to `AosPsaAddKey`, `AosPsaRemoveKey`.

File: src/crypto/crypto.cpp [hunk 13]
```diff
@@ -8,6 +8,58 @@
...
+    mbedtls_pk_context    key;
+
+    InitializeCSR(csr, key);
+
+    auto cleanup = [&]() {
+        mbedtls_x509write_csr_free(&csr);
+        mbedtls_pk_free(&key);
+    };
+
+    mMutex.Lock();
```
Comment (al1img) [inline src/crypto/crypto.cpp:85]: If you guard Add/Remove key, use static mutex in driver itself.

File: src/crypto/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,73 @@
+# #
```
Comment (al1img) [inline src/crypto/CMakeLists.txt:1]: It should be part of common lib. Mbedtls should be linked always for tests.

File: src/crypto/mbedtls/psa_crypto_driver_wrappers.c [hunk 15]
```diff
@@ -0,0 +1,2887 @@
+/*
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/crypto/mbedtls/psa_crypto_driver_wrappers.c]: Done

File: include/aos/crypto/mbedtls/drivers/aos/driver_wrap.h (hunk 1 above)
Comment (MykolaSolyanko) [reply include/aos/crypto/mbedtls/drivers/aos/driver_wrap.h]: Done

File: include/aos/crypto/mbedtls/drivers/aos/driver_wrap.h (hunk 2 above)
Comment (MykolaSolyanko) [reply include/aos/crypto/mbedtls/drivers/aos/driver_wrap.h]: Done

File: include/aos/crypto/mbedtls/drivers/aos/driver_wrap.h (hunk 3 above)
Comment (MykolaSolyanko) [reply include/aos/crypto/mbedtls/drivers/aos/driver_wrap.h]: Done

File: src/crypto/mbedtls/drivers/aos/driver_wrap.cpp (hunk 4 above)
Comment (MykolaSolyanko) [reply src/crypto/mbedtls/drivers/aos/driver_wrap.cpp]: Done

File: src/crypto/mbedtls/drivers/aos/driver_wrap.cpp (hunk 5 above)
Comment (MykolaSolyanko) [reply src/crypto/mbedtls/drivers/aos/driver_wrap.cpp]: Done

File: src/crypto/mbedtls/drivers/aos/driver_wrap.cpp (hunk 6 above)
Comment (MykolaSolyanko) [reply src/crypto/mbedtls/drivers/aos/driver_wrap.cpp]: Done

File: src/crypto/mbedtls/drivers/aos/driver_wrap.cpp (hunk 7 above)
Comment (MykolaSolyanko) [reply src/crypto/mbedtls/drivers/aos/driver_wrap.cpp]: Done

File: src/crypto/mbedtls/drivers/aos/driver_wrap.cpp (hunk 9 above)
Comment (MykolaSolyanko) [reply src/crypto/mbedtls/drivers/aos/driver_wrap.cpp]: Done

File: include/aos/crypto/crypto.hpp (hunk 10 above)
Comment (MykolaSolyanko) [reply include/aos/crypto/crypto.hpp]: Done

File: include/aos/crypto/mbedtls/drivers/aos/driver_wrap.h (hunk 12 above)
Comment (MykolaSolyanko) [reply include/aos/crypto/mbedtls/drivers/aos/driver_wrap.h]: Done

File: src/crypto/crypto.cpp (hunk 13 above)
Comment (MykolaSolyanko) [reply src/crypto/crypto.cpp]: Done

File: src/crypto/CMakeLists.txt (hunk 14 above)
Comment (MykolaSolyanko) [reply src/crypto/CMakeLists.txt]: Done

File: src/common/CMakeLists.txt [hunk 16]
```diff
@@ -20,24 +22,69 @@ endif()
...
 
+target_link_libraries(${TARGET} PUBLIC mbedcrypto mbedx509 mbedtls)
+
 # ######################################################################################################################
 # Install
 # ######################################################################################################################
 
 install(TARGETS ${TARGET} ARCHIVE DESTINATION lib)
 
+if(NOT USE_INTERNAL_MBEDTLS)
```
Comment (al1img) [inline src/common/CMakeLists.txt:79]: I guess it is not required.

File: src/common/CMakeLists.txt [hunk 17]
```diff
@@ -7,6 +7,8 @@
 
 set(TARGET aoscommoncpp)
 
+option(USE_INTERNAL_MBEDTLS "Use internal mbedtls library" OFF)
```
Comment (al1img) [inline src/common/CMakeLists.txt:10]: `aoscommoncpp` should be built without link with mbedtls. It is up to client (zephyr app) to provide mbedtls. So, it should be moved to tests cmake I guess.

Comment (al1img) [inline src/common/CMakeLists.txt:10]: Use options header as following:

```cmake
# ######################################################################################################################
# Options
# ######################################################################################################################

option(WITH_MBEDTLS "build with mbedTLS" ON)

message(STATUS "WITH_MBEDTLS                  = ${WITH_MBEDTLS}")
message(STATUS)
```

WITH_MBEDTLS - means build together with mbedtls

File: src/common/CMakeLists.txt [hunk 18]
```diff
@@ -20,25 +22,73 @@ endif()
...
+# ######################################################################################################################
+# Fetch and setup MbedTLS
+# ######################################################################################################################
+
+if(NOT USE_INTERNAL_MBEDTLS)
+    include(FetchContent)
+
+    set(FETCHCONTENT_QUIET FALSE)
+
+    FetchContent_Declare(
```
Comment (al1img) [inline src/common/CMakeLists.txt:39]: Use cmake-format

File: src/common/CMakeLists.txt [hunk 19]
```diff
@@ -20,25 +22,73 @@ endif()
...
+    uuid.cpp
+    pkcs11/pkcs11.cpp
+    pkcs11/privatekey.cpp
+    cryptoutils.cpp
+    crypto/mbedtls/crypto.cpp
+    crypto/mbedtls/drivers/aos/driverwrap.cpp)
 
 target_sources(${TARGET} PUBLIC ${CMAKE_CURRENT_BINARY_DIR}/version.hpp)
 
+target_link_libraries(${TARGET} PUBLIC mbedcrypto mbedx509 mbedtls)
```
Comment (al1img) [inline src/common/CMakeLists.txt:75]: Link only if WITH_MBEDTLS

File: tests/common/src/crypto_test.cpp [hunk 20]
```diff
@@ -0,0 +1,381 @@
+/*
```
Comment (mykola-kobets-epam) [inline tests/common/src/crypto_test.cpp:1]: there are two crypto_test.cpp files in the review.

File: src/common/crypto/mbedtls/crypto.cpp [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+        return AOS_ERROR_WRAP(ret);
+    }
+
+    mbedtls_x509_crt* currentCrt = &crt;
+
+    while (currentCrt != nullptr) {
+        aos::crypto::x509::Certificate cert;
+
+        auto err = ParseX509Certs(currentCrt, cert);
+        if (err != aos::ErrorEnum::eNone) {
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/common/crypto/mbedtls/crypto.cpp:136]: for me it looks like using `ptr != nullptr` or `!ptr`, but we don't have a rule that we should use to check on error like this. But ok i will update

File: include/aos/common/crypto/mbedtls/drivers/aos/driverwrap.hpp [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+aos::RetWithError<psa_key_id_t> AosPsaAddKey(const aos::crypto::PrivateKeyItf& mPrivKey);
+
+/**
+ * @brief Remove key from the list of builtin keys
+ *
+ * @param keyId Key ID
+ */
+void AosPsaRemoveKey(psa_key_id_t keyId);
+
+#endif
```
Comment (al1img) [inline include/aos/common/crypto/mbedtls/drivers/aos/driverwrap.hpp:31]: New line at the end.

File: src/common/crypto/mbedtls/drivers/aos/driverwrap.cpp [hunk 23]
```diff
@@ -0,0 +1,210 @@
...
+#include <mbedtls/pk.h>
+#include <mbedtls/rsa.h>
+
+#include "aos/common/crypto/mbedtls/drivers/aos/driverwrap.hpp"
+#include "aos/common/tools/array.hpp"
+#include "aos/common/tools/thread.hpp"
+
+/**
+ * Key description
+ *
```
Comment (al1img) [inline src/common/crypto/mbedtls/drivers/aos/driverwrap.cpp:21]: Remove extra line.

File: src/common/crypto/mbedtls/drivers/aos/driverwrap.cpp [hunk 24]
```diff
@@ -0,0 +1,210 @@
...
+#include <mbedtls/pem.h>
+#include <mbedtls/pk.h>
+#include <mbedtls/rsa.h>
+
+#include "aos/common/crypto/mbedtls/drivers/aos/driverwrap.hpp"
+#include "aos/common/tools/array.hpp"
+#include "aos/common/tools/thread.hpp"
+
+/**
+ * Key description
```
Comment (al1img) [inline src/common/crypto/mbedtls/drivers/aos/driverwrap.cpp:20]: Period at the end.

File: src/common/crypto/mbedtls/drivers/aos/driverwrap.cpp [hunk 25]
```diff
@@ -0,0 +1,210 @@
...
+ *
+ */
+struct KeyDescription {
+    /**
+     * Key ID
```
Comment (al1img) [inline src/common/crypto/mbedtls/drivers/aos/driverwrap.cpp:25]: Period at the end of comment.

File: src/common/crypto/mbedtls/drivers/aos/driverwrap.cpp [hunk 26]
```diff
@@ -0,0 +1,210 @@
...
+        if (key.mKeyId == keyId) {
+            sBuiltinKeys.Remove(&key);
+
+            psa_destroy_key(MBEDTLS_SVC_KEY_ID_GET_KEY_ID(keyId));
+
+            break;
+        }
+    }
+}
+
```
Comment (al1img) [inline src/common/crypto/mbedtls/drivers/aos/driverwrap.cpp:259]: Add header :
```c++
/***********************************************************************************************************************
 * PSA driver wrappers
 **********************************************************************************************************************/
```

File: src/common/crypto/mbedtls/drivers/aos/driverwrap.cpp [hunk 27]
```diff
@@ -0,0 +1,210 @@
...
+#include <mbedtls/base64.h>
+#include <mbedtls/pem.h>
+#include <mbedtls/pk.h>
+#include <mbedtls/rsa.h>
+
+#include "aos/common/crypto/mbedtls/drivers/aos/driverwrap.hpp"
+#include "aos/common/tools/array.hpp"
+#include "aos/common/tools/thread.hpp"
+
+/**
```
Comment (al1img) [inline src/common/crypto/mbedtls/drivers/aos/driverwrap.cpp:19]: Add header:
```c++
/***********************************************************************************************************************
 * Structs
 **********************************************************************************************************************/
```

File: include/aos/common/crypto.hpp [hunk 28]
```diff
@@ -150,7 +150,7 @@ class PublicKeyItf {
      *
      * @return KeyType,
      */
-    virtual KeyType GetKeyType() const = 0;
+    virtual KeyTypeEnum GetKeyType() const = 0;
```
Comment (al1img) [inline include/aos/common/crypto.hpp:158]: Revert this changes. Use GetValue method of KeyType in switch.

File: tests/common/src/crypto_test.cpp [hunk 29]
```diff
@@ -0,0 +1,677 @@
...
+#include <mbedtls/oid.h>
+#include <mbedtls/pk.h>
+#include <mbedtls/x509_crt.h>
+
+#include "aos/common/crypto/mbedtls/crypto.hpp"
+
+/***********************************************************************************************************************
+ * Static
+ **********************************************************************************************************************/
+static std::pair<aos::Error, std::vector<unsigned char>> GenerateRSAPrivateKey()
```
Comment (al1img) [inline tests/common/src/crypto_test.cpp:25]: Empty line after header.

File: include/aos/common/crypto/mbedtls/crypto.hpp [hunk 30]
```diff
@@ -0,0 +1,91 @@
...
+
+#include <mbedtls/ctr_drbg.h>
+#include <mbedtls/entropy.h>
+#include <mbedtls/x509_crt.h>
+#include <mbedtls/x509_csr.h>
+#include <psa/crypto_types.h>
+
+#include "aos/common/crypto.hpp"
+
+class Crypto : public aos::crypto::x509::ProviderItf {
```
[the remark this answers is not in the document]
Comment (al1img) [reply include/aos/common/crypto/mbedtls/crypto.hpp:25]: Interface should be in crypto as well.
Rename crypto to `MbedTLSCryptoProvider`.
files should be `cryptoprovider.hpp`, `cryptoprovider.cpp` accordingly.

File: include/aos/common/crypto/mbedtls/psa_crypto_driver_wrappers.c [hunk 31]
```diff
@@ -0,0 +1,2889 @@
+/*
```
Comment (mykola-kobets-epam) [inline include/aos/common/crypto/mbedtls/psa_crypto_driver_wrappers.c:1]: *.c, *.cpp files should not be added to include directory

File: src/common/crypto/mbedtls/crypto.cpp [hunk 32]
```diff
@@ -80,6 +80,31 @@ aos::Error Crypto::DERToX509Cert(const aos::Array<uint8_t>& derBlob, aos::crypto
...
+
+    result.Resize(result.MaxSize());
+    uint8_t* start = result.Get();
+    uint8_t* p     = start + result.Size();
+
+    ret = mbedtls_x509_write_names(&p, start, dn);
+    if (ret < 0) {
+        return AOS_ERROR_WRAP(ret);
+    }
+
```
Comment (mykola-kobets-epam) [inline src/common/crypto/mbedtls/crypto.cpp:199]: dn should be freed

File: tests/common/src/crypto_test.cpp [hunk 33]
```diff
@@ -0,0 +1,677 @@
...
+{
+    mbedtls_pk_context       pk;
```
Comment (mykola-kobets-epam) [inline tests/common/src/crypto_test.cpp:27]: For the tests we could use generated keys. But Im ok with this approach too.

File: src/common/CMakeLists.txt [hunk 34]
```diff
@@ -25,14 +30,62 @@ execute_process(
...
+if(WITH_MBEDTLS)
+    include(FetchContent)
+
+    set(FETCHCONTENT_QUIET FALSE)
+
+    FetchContent_Declare(
+        mbedtls
+        GIT_REPOSITORY https://github.com/Mbed-TLS/mbedtls.git
+        GIT_TAG v3.4.1
+        GIT_PROGRESS TRUE
```
Comment (mykola-kobets-epam) [inline src/common/CMakeLists.txt:46]: GIT_SHALLOW flag can speed up mbedtls download

File: include/aos/common/crypto/mbedtls/psa_crypto_driver_wrappers.c (hunk 31 above)
Comment (MykolaSolyanko) [reply include/aos/common/crypto/mbedtls/psa_crypto_driver_wrappers.c:1]: @al1img  asked to do it

File: tests/common/src/crypto_test.cpp (hunk 33 above)
Comment (MykolaSolyanko) [reply tests/common/src/crypto_test.cpp:27]: this security policy required
https://github.com/aoscloud/aos_core_lib_cpp/pull/59#discussion_r1466101084

File: include/aos/common/crypto/mbedtls/psa_crypto_driver_wrappers.c (hunk 31 above)
Comment (MykolaSolyanko) [reply include/aos/common/crypto/mbedtls/psa_crypto_driver_wrappers.c:1]: https://github.com/aoscloud/aos_core_lib_cpp/pull/59#discussion_r1466009928

File: .github/workflows/check_format.yaml [hunk 35]
```diff
@@ -23,6 +23,7 @@ jobs:
         uses: jidicula/clang-format-action@v4.11.0
         with:
           clang-format-version: "15"
+          exclude-regex: 'driver\.h|mbedtls_config\.h|psa_crypto_driver_wrappers\.c'
```
Comment (al1img) [inline .github/workflows/check_format.yaml:26]: driver.h could be formatted with our format.

File: include/aos/common/crypto/mbedtls/psa_crypto_driver_wrappers.c (hunk 31 above)
Comment (al1img) [reply include/aos/common/crypto/mbedtls/psa_crypto_driver_wrappers.c:1]: I meant `src` folder. Sorry for that.
